### PR TITLE
[Balance Druid] Fix to make moons available casts correct in CastEfficiency module

### PR DESCRIPTION
**Pull request by Iskalla**

### CODE REVIEW

File: src/Parser/BalanceDruid/Modules/Features/CastEfficiency.js [hunk 1]
```diff
@@ -13,7 +12,8 @@ class CastEfficiency extends CoreCastEfficiency {
       spell: SPELLS.NEW_MOON,
       category: CastEfficiency.SPELL_CATEGORIES.ROTATIONAL,
       getCooldown: (haste, combatant) => {
-        const cd = (combatant.owner.fightDuration / 1000) / (((combatant.owner.fightDuration / 1000) - 2) / 45);
+        const availableCasts = combatant.owner.modules.newmoon.nmAvailableCasts;
+        const cd = ((combatant.owner.fightDuration / 1000) / availableCasts);
```
Comment (MartijnHols) [inline src/Parser/BalanceDruid/Modules/Features/CastEfficiency.js]: This var can be omitted and its value returned immediately.

File: src/Parser/BalanceDruid/Modules/Features/CastEfficiency.js [hunk 2]
```diff
@@ -22,7 +22,8 @@ class CastEfficiency extends CoreCastEfficiency {
       spell: SPELLS.HALF_MOON,
       category: CastEfficiency.SPELL_CATEGORIES.ROTATIONAL,
       getCooldown: (haste, combatant) => {
-        const cd = (combatant.owner.fightDuration / 1000) / (((combatant.owner.fightDuration / 1000) - 17) / 45);
+        const availableCasts = combatant.owner.modules.halfmoon.hmAvailableCasts;
+        const cd = (combatant.owner.fightDuration / 1000) / availableCasts;
```
Comment (MartijnHols) [inline src/Parser/BalanceDruid/Modules/Features/CastEfficiency.js]: This var can be omitted and its value returned immediately.

File: src/Parser/BalanceDruid/Modules/Features/CastEfficiency.js [hunk 3]
```diff
@@ -31,9 +32,9 @@ class CastEfficiency extends CoreCastEfficiency {
       spell: SPELLS.FULL_MOON,
       category: CastEfficiency.SPELL_CATEGORIES.ROTATIONAL,
       getCooldown: (haste, combatant) => {
-        const hasMooncloak = combatant.hasBack(ITEMS.RADIANT_MOONLIGHT.id);
-        const cd = (combatant.owner.fightDuration / 1000) / (((combatant.owner.fightDuration / 1000) - 32) / 45);
-        return hasMooncloak ? cd * 2 : cd;
+        const availableCasts = combatant.owner.modules.fullmoon.fmAvailableCasts;
+        const cd = (combatant.owner.fightDuration / 1000) / availableCasts;
```
Comment (MartijnHols) [inline src/Parser/BalanceDruid/Modules/Features/CastEfficiency.js]: This var can be omitted and its value returned immediately.

File: src/Parser/BalanceDruid/Modules/Spells/FullMoon.js [hunk 4]
```diff
@@ -65,17 +67,17 @@ class FullMoon extends Module {
     const offSet = this.firstMoonTime + 15;
     const totalFromCD = ((this.owner.fightDuration / 1000) - offSet) / 15;
     const eachMoon = Math.floor(totalFromCD / 3);
-    let fmAvailableCasts = eachMoon + 1;
+    this.fmAvailableCasts = eachMoon + 1;
 
     const extraMoons = ((totalFromCD / 3) - eachMoon) * 3;
-    if (extraMoons > this.fullMoonOrder) fmAvailableCasts += 1;
+    if (extraMoons > this.fullMoonOrder) this.fmAvailableCasts += 1;
```
Comment (MartijnHols) [inline src/Parser/BalanceDruid/Modules/Spells/FullMoon.js]: missing braces

File: src/Parser/BalanceDruid/Modules/Spells/FullMoon.js [hunk 5]
```diff
@@ -40,14 +42,14 @@ class FullMoon extends Module {
...
```
Comment (MartijnHols) [inline src/Parser/BalanceDruid/Modules/Spells/FullMoon.js]: missing braces

File: src/Parser/BalanceDruid/Modules/Spells/FullMoon.js [hunk 6]
```diff
@@ -40,14 +42,14 @@ class FullMoon extends Module {
     const offSet = this.firstMoonTime + 15;
     const totalFromCD = ((this.owner.fightDuration / 1000) - offSet) / 15;
     const eachMoon = Math.floor(totalFromCD / 3);
-    let fmAvailableCasts = eachMoon + 1;
+    this.fmAvailableCasts = eachMoon + 1;
```
Comment (MartijnHols) [inline src/Parser/BalanceDruid/Modules/Spells/FullMoon.js]: Setting `fmAvailableCasts` in suggestions is fragile as it being called is not guaranteed. A better solution would be to add a getter for `fmAvailableCasts` that calculates it whenever its called.

File: src/Parser/BalanceDruid/Modules/Spells/HalfMoon.js [hunk 7]
```diff
@@ -40,14 +42,14 @@ class HalfMoon extends Module {
     const offSet = this.firstMoonTime + 15;
     const totalFromCD = ((this.owner.fightDuration / 1000) - offSet) / 15;
     const eachMoon = Math.floor(totalFromCD / 3);
-    let hmAvailableCasts = eachMoon + 1;
+    this.hmAvailableCasts = eachMoon + 1;
```
Comment (MartijnHols) [inline src/Parser/BalanceDruid/Modules/Spells/HalfMoon.js]: See fmAvailableCasts comment

File: src/Parser/BalanceDruid/Modules/Spells/NewMoon.js [hunk 8]
```diff
@@ -40,13 +46,13 @@ class NewMoon extends Module {
     const offSet = this.firstMoonTime + 15;
     const totalFromCD = ((this.owner.fightDuration / 1000) - offSet) / 15;
     const eachMoon = Math.floor(totalFromCD / 3);
-    let nmAvailableCasts = eachMoon + 1;
+    this.nmAvailableCasts = eachMoon + 1;
```
Comment (MartijnHols) [inline src/Parser/BalanceDruid/Modules/Spells/NewMoon.js]: See fmAvailableCasts comment

File: src/Parser/BalanceDruid/Modules/Spells/FullMoon.js [hunk 9]
```diff
@@ -15,7 +15,17 @@ class FullMoon extends Module {
   firstMoonCast = false;
   orderFound = false;
 
-  fmAvailableCasts = 0;
+  getfmAvailableCasts() {
```
Comment (MartijnHols) [inline src/Parser/BalanceDruid/Modules/Spells/FullMoon.js]: `get fmAvailableCasts()` will allow you to use this as a property (this.fmAvailableCasts)

File: src/Parser/BalanceDruid/Modules/Spells/NewMoon.js [hunk 10]
```diff
@@ -67,20 +71,13 @@ class NewMoon extends Module {
...
-    const offSet = this.firstMoonTime + 15;
-    const totalFromCD = ((this.owner.fightDuration / 1000) - offSet) / 15;
-    const eachMoon = Math.floor(totalFromCD / 3);
-    this.nmAvailableCasts = eachMoon + 1;
-
-    const extraMoons = ((totalFromCD / 3) - eachMoon) * 3;
-    if (extraMoons > this.newMoonOrder) this.nmAvailableCasts += 1;
-
     const nmCasted = abilityTracker.getAbility(SPELLS.NEW_MOON.id).casts;
+    const nmAvailable = this.getnmAvailableCasts();
```
Comment (MartijnHols) [inline src/Parser/BalanceDruid/Modules/Spells/NewMoon.js]: This temp property can be omitted if you turn `getnmAvailableCasts` into a getter

File: src/Parser/BalanceDruid/Modules/Features/CastEfficiency.js [hunk 11]
```diff
@@ -13,27 +12,26 @@ class CastEfficiency extends CoreCastEfficiency {
       spell: SPELLS.NEW_MOON,
       category: CastEfficiency.SPELL_CATEGORIES.ROTATIONAL,
       getCooldown: (haste, combatant) => {
-        const cd = (combatant.owner.fightDuration / 1000) / (((combatant.owner.fightDuration / 1000) - 2) / 45);
-        return cd;
+        const availableCasts = combatant.owner.modules.newmoon.nmAvailableCasts;
+        return ((combatant.owner.fightDuration / 1000) / availableCasts);
```
Comment (MartijnHols) [inline src/Parser/BalanceDruid/Modules/Features/CastEfficiency.js]: One set of parenthesis can be omitted 😅

File: src/Parser/BalanceDruid/Modules/Spells/FullMoon.js [hunk 12]
```diff
@@ -15,6 +15,18 @@ class FullMoon extends Module {
...
   orderFound = false;
 
+  get fmAvailableCasts() {
+    const offSet = this.firstMoonTime + 15;
+    const totalFromCD = ((this.owner.fightDuration / 1000) - offSet) / 15;
+    const eachMoon = Math.floor(totalFromCD / 3);
+    let fmAvailableCasts = eachMoon + 1;
+
+    const extraMoons = ((totalFromCD / 3) - eachMoon) * 3;
+    if (extraMoons > this.fullMoonOrder) { fmAvailableCasts += 1; }
```
Comment (MartijnHols) [inline src/Parser/BalanceDruid/Modules/Spells/FullMoon.js]: This should be multiline.

File: src/Parser/BalanceDruid/Modules/Features/CastEfficiency.js (hunk 11 above)
Comment (Iskalla) [reply src/Parser/BalanceDruid/Modules/Features/CastEfficiency.js]: You are having fun today picking apart my code, hm? xDD

Comment (MartijnHols) [reply src/Parser/BalanceDruid/Modules/Features/CastEfficiency.js]: just being thorough, approved the changes though so you can merge without fixing if you so desire.

**Comment by MartijnHols** [comment]
Thanks!

**Pull request merged**